Mockup of a navbar with a collapsible sidebar in a narrow view width

dnswrsrx opened the pull request:
Navbar mockup from #18 with some additional goodies and tweaks.
[Site Preview](https://ui-project-seven.vercel.app/).

As a collapsible sidebar could be great for narrow view widths (mobile) and the navbar could hold a button to toggle this collapsible sidebar, this PR may include this feature. I can isolate the collapsible navbar into its own PR if there's more discussion to be had about sidebar(s) in general.

Implemented a user icon-ish dropdown in the navbar (yay HeadlessUI). If there's no user in scope (not logged in), the dropdown will contain links to signing up and logging in. If there's a user in scope (logged in), it will show a different list that is relevant to the user, _e.g._ profile, settings.

Since user icons could be used in the navbar, took the opportunity to quickly refactor the now deprecated `UserIcon` component. Isolated the Avatar image portion to its own component, while moving all the username and display name handling out of the `UserIcon` component into the sidebar component where it was only used in this manner. Hopefully the `Avatar` icon is a decent abstraction as it can be used in several components in a similar way.

Made some additional minor tweaks to the appearance and spacing of various components.

dnswrsrx committed:
Mock up a navbar based on the proposed design

Logo with instance name on the left

Search bar (just an input field currently) in the center

A container on the right - thinking of something user/account based,
notifications, and a button to show/hide sidebar on smaller view width
e.g. mobile device.

dnswrsrx committed:
Implement collapsing sidebar with a button for narrow view width

Making use of absolute positioning to "stick" the sidebar to the right

Using width 0 and 80% to "collapse" and "show" sidebar respectively
Transitions to make the collapsing and showing seem "dynamic"

Making the toggle button a client component to use state to determine
when to show/hide the sidebar as well as which icons to use.

Applying some filters to the main article element to make it appear as
if it's "fading off". Created a CSS class for this, and toggling this
class on the main article element.

Use lg breakpoint for the width setting on the main article and sidebar.
Allows the main article to take the full width, as it should, when it's
a narrow view e.g. on mobile.

dnswrsrx committed:
Some tweaks to space and text in sidebar

Move bottom margin to parent as we're tweaking space between parents

Add margin between admins header, and add some gap between each admin in
the list.

dnswrsrx committed:
Use instance name in the logo's alt text

Since we have the instance info in context

dnswrsrx committed:
More distinction to the sidebar from background in narrow view width

Add a slight left border and increase overall padding in narrow view
width. Hopefully helps visually separate the sidebar better.

dnswrsrx committed:
Abstract avatar into its component while removing the UserIcon component

The avatar could benefit being its own component as it's used in several
other components, and currently, abstraction for this seems to be fine.

The previous UserIcon component does a bit more than just icons
delivering other username / display name things. It seemed to be
specific to the sidebar. So moved the username / display name code to
the sidebar while using the new Avatar component, and removed this
UserIcon component.

dnswrsrx committed:
Use search button instead of search bar in narrow view width

As per the proposed design for mobile view.
https://design.penpot.app/#/view/c65f3168-697c-808c-8002-e05c7100c00a?page-id=c65f3168-697c-808c-8002-e05c7100c00b&section=interactions&index=0&share-id=3d915f59-a8d9-81d6-8002-e0693b579375

dnswrsrx committed:
Implement a user icon dropdown in the navbar

If no user in scope, dropdown includes links to signing up or logging
in.

If there's a user in scope, will modify the list of links to include
relevant things to user, such as profile, settings.

lionirdeadman reviewed (changes requested):
- [ ] If you open the sidebar in a small viewport and then make the viewport big again, the blur does not go away
- [x] The sidebar could be full width when on mobile. 
- [x] This reverts the colour change of the navbar introduced in #37, was that intentional?
- [x] We should standardize if we use CamelCase or lowercase for component names
- [x] Wondering if @sunaurus already has the user dropdown or something equivalent in their work on auth/login

dnswrsrx committed:
Fix the navbar colour reversion from the rebase on PR #37

dnswrsrx committed:
useEffect in SidebarToggle should have the dependency specified

This effect is dependent on the expanded variable changing. It was
working fine because the component was re-rendering, but shouldn't rely
on the re-rendering for the effect to run.

dnswrsrx committed:
Show instance name in navbar in mobile view as well

Since the search bar becomes a button in mobile view, there's enough
space for the name.

dnswrsrx committed:
Different side bar widths for narrow view widths

Use full width when small, otherwise 4/5 width for medium.

dnswrsrx committed:
Use CamelCase for navbar and avatar component directories

CamelCase for component names is a pretty common convention. If the
component directory contains an index file, I think it makes sense to
use CamelCase, even though we're importing content from the file.

dnswrsrx commented:
> If you open the sidebar in a small viewport and then make the viewport big again, the blur does not go away

Is this still happening? Try a hard refresh in case there's stale CSS (`ctrl/cmd + shift + r`). From my testing on Firefox and Chromium, it seems to be fine for me: the blur disappears when making viewport wide again with the sidebar opened.

> The sidebar could be full width when on mobile.

I made the sidebar take the full width at the narrowest break point, and 80% width at the next wider break point.

> This reverts the colour change of the navbar introduced in https://github.com/beehaw-org/ui-project/pull/37, was that intentional?

Accidentally overwrote this when rebasing.

> We should standardize if we use CamelCase or lowercase for component names

I assume you're referring to the directory of the component names - in this case `Navbar` and `Avatar`? If so, I think CamelCase makes sense, especially if we're using the `index.` file where the compiler allows us to import the directory.

Within the code however, I believe we are using CamelCase for all the components so far?

> Wondering if @sunaurus already has the user dropdown or something equivalent in their work on auth/login

If this is the case I'm happy to drop this for this PR or it can be overwritten in the following PR.

lionirdeadman commented:
> Is this still happening? Try a hard refresh in case there's stale CSS (ctrl/cmd + shift + r). From my testing on Firefox and Chromium, it seems to be fine for me: the blur disappears when making viewport wide again with the sidebar opened.

Yes. The way I'm testing is that I open the dev tools, make it small enough to show the sidebar button, click on it and then make the viewport wide again. It does correctly go on the side but the contents are still blurred.

> I assume you're referring to the directory of the component names - in this case Navbar and Avatar? If so, I think CamelCase makes sense, especially if we're using the index. file where the compiler allows us to import the directory.

Yeah, the folder names.

Rest LGTM, thank you

dnswrsrx commented:
> Yes. The way I'm testing is that I open the dev tools, make it small enough to show the sidebar button, click on it and then make the viewport wide again. It does correctly go on the side but the contents are still blurred.

I'm still not able to replicate this. Here are some screenshots of the testing I did on Firefox, I think using the way you're testing?

Here I set the width to 800px which shows the button to bring up the sidebar. I click it and the sidebar is shown with the blur occurring.

<img width="1227" alt="medium_withblur" src="https://github.com/beehaw-org/ui-project/assets/35418785/674caef6-eadc-4d75-b58c-72ebf7a8dac4">

Now, I set the width to 1100px without closing the sidebar, which should always show the sidebar and hide that button. At this point, the blur should also disappear even though the sidebar wasn't closed prior - this is the case for me as per the screenshot below. I assume this isn't the case for you? Note the `dim` class on the `article` element present between the two screenshots: it has styling for the blur, but the blur should only apply on the narrow screen widths.

<img width="1264" alt="wider_blur_off" src="https://github.com/beehaw-org/ui-project/assets/35418785/cb8c2c23-c187-4f6c-b01d-9c872108906e">

sunaurus commented:
>> Wondering if @sunaurus already has the user dropdown or something equivalent in their work on auth/login

>If this is the case I'm happy to drop this for this PR or it can be overwritten in the following PR.

I will rebase my branch on to this one actually 👍
